# Часть 1. Подключение и обзор данных

**Проект:** анализ Telegram-канала @krylikryli (Крыли) за 12 месяцев.

## Инструкция по запуску

### 1. Зависимости
```powershell
pip install pandas sqlalchemy psycopg2-binary seaborn matplotlib scipy statsmodels
```

### 2. Postgres должен быть поднят
```powershell
docker ps   # контейнер kryli_postgres должен быть Up
```

### 3. Запускать ячейки сверху вниз
Kernel — интерпретатор из `.venv`.

### 4. Параметры подключения
Если логин/пароль/порт другие — правь в ячейке `PARS`.

---

**Что в этой части:** подключение к БД, выгрузка витрин и `cleansed_posts`, общий обзор (период, диапазоны, топ-посты).

In [1]:
# ============================================================
# ШАГ 1. ПАРАМЕТРЫ ПОДКЛЮЧЕНИЯ
# ============================================================
PG_USER = "kryli"
PG_PASS = "kryli_pass"
PG_HOST = "localhost"   # Postgres слушает порт на хосте
PG_PORT = "5432"
PG_DB   = "kryli"

from sqlalchemy import create_engine
engine = create_engine(
    f"postgresql+psycopg2://{PG_USER}:{PG_PASS}@{PG_HOST}:{PG_PORT}/{PG_DB}"
)
print("Подключение готово:", engine.url.render_as_string(hide_password=True))

Подключение готово: postgresql+psycopg2://kryli:***@localhost:5432/kryli


In [2]:
# ============================================================
# ШАГ 2. ИМПОРТЫ И НАСТРОЙКА ВИЗУАЛИЗАЦИИ
# ============================================================
# Настройки графиков повторяются в части 2 — так каждая часть автономна.
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_theme(style="whitegrid", palette="muted")
plt.rcParams["figure.figsize"] = (10, 5)
plt.rcParams["axes.titlesize"] = 13
plt.rcParams["figure.dpi"] = 110
pd.set_option("display.max_columns", 50)
pd.set_option("display.width", 160)
print("Библиотеки загружены")

Библиотеки загружены


In [3]:
# ============================================================
# ШАГ 3. ВЫГРУЗКА ВИТРИН И CLEANSED
# ============================================================
m_hour      = pd.read_sql("SELECT * FROM mart_by_hour",      engine)
m_weekday   = pd.read_sql("SELECT * FROM mart_by_weekday",   engine)
m_type      = pd.read_sql("SELECT * FROM mart_by_content_type", engine)
m_flags     = pd.read_sql("SELECT * FROM mart_by_text_flags", engine)
m_interval  = pd.read_sql("SELECT * FROM mart_by_interval",  engine)
m_length    = pd.read_sql("SELECT * FROM mart_by_text_length", engine)
m_combo     = pd.read_sql("SELECT * FROM mart_by_combination", engine)

df = pd.read_sql("SELECT * FROM cleansed_posts", engine)
df["date_time"] = pd.to_datetime(df["date_time"], utc=True)

print(f"cleansed_posts: {df.shape[0]} строк, {df.shape[1]} колонок")
print(f"mart_by_hour: {m_hour.shape[0]} строк")
print(f"mart_by_combination: {m_combo.shape[0]} строк")
df.head(3)

cleansed_posts: 1480 строк, 26 колонок
mart_by_hour: 24 строк
mart_by_combination: 8 строк


,post_id,date_time,text,views,reactions_total,reactions_breakdown,forwards,replies,is_pinned,is_forwarded,hashtags,media_type,text_length,hour,weekday,is_recent,reactions_per_view,replies_per_view,forwards_per_view,er,time_since_prev_post,is_personal_heur,er_norm,length_tertile,length_bucket,is_holiday
0,3565,2026-06-18 13:04:16+00:00,,4254,2,"{'ReactionPaid': 1, 'ReactionCustomEmoji': 1}",10,17,False,False,[],photo,0,13.0,4.0,False,0.000470,0.003996,0.002351,0.006817,0.00,False,1.0433,1,short,False
1,3564,2026-06-18 12:40:59+00:00,,1695,3,{'ReactionCustomEmoji': 3},7,0,False,True,[],photo,0,12.0,4.0,False,0.001770,0.000000,0.004130,0.005900,14.56,False,0.9030,1,short,False
2,3562,2026-06-17 11:40:55+00:00,,1995,12,"{'ReactionPaid': 1, 'ReactionCustomEmoji': 11}",13,19,False,False,[],document,0,11.0,3.0,False,0.006015,0.009524,0.006516,0.022055,4.20,False,3.3754,1,short,False


In [4]:
# ============================================================
# ШАГ 4. ОБЩАЯ СТАТИСТИКА ПО ПЕРИОДУ И RANGE
# ============================================================
print("Период данных:", df["date_time"].min(), "—", df["date_time"].max())
print("Всего постов:", len(df))
print("Свежих (< 7 дней):", int(df["is_recent"].sum()))
print("Праздничных:", int(df["is_holiday"].sum()))
print("Личных (эвристика):", int(df["is_personal_heur"].sum()))
print("Закреплённых:", int(df["is_pinned"].sum()))
print("Форвардов:", int(df["is_forwarded"].sum()))

df[["views", "reactions_total", "replies", "forwards", "er", "er_norm"]].describe().round(4)

Период данных: 2025-10-13 19:19:27+00:00 — 2026-10-08 11:50:23+00:00
Всего постов: 1480
Свежих (< 7 дней): 45
Праздничных: 45
Личных (эвристика): 45
Закреплённых: 1
Форвардов: 294


,views,reactions_total,replies,forwards,er,er_norm
count,1480.0000,1480.0000,1480.0000,1480.0000,1480.0000,1480.0000
mean,2285.8351,5.6730,7.2993,8.1676,0.0093,1.4220
std,949.2594,10.9517,17.6130,12.3313,0.0102,1.5578
min,344.0000,0.0000,0.0000,0.0000,0.0000,0.0000
25%,1665.5000,0.0000,0.0000,2.0000,0.0030,0.4662
50%,2026.0000,2.0000,2.0000,4.0000,0.0065,1.0000
75%,2799.0000,7.0000,8.0000,10.0000,0.0119,1.8139
max,11323.0000,140.0000,307.0000,140.0000,0.1074,16.4408


In [5]:
# ============================================================
# ШАГ 5. TOP-5 ПОСТОВ ПО ПРОСМОТРАМ И ПО ER
# ============================================================
cols = ["post_id", "date_time", "views", "reactions_total", "replies", "forwards", "er", "er_norm"]

print("=== TOP-5 по просмотрам ===")
display(df.nlargest(5, "views")[cols])

print("=== TOP-5 по ER ===")
display(df.nlargest(5, "er")[cols])

=== TOP-5 по просмотрам ===


,post_id,date_time,views,reactions_total,replies,forwards,er,er_norm
126,3691,2026-07-14 17:33:11+00:00,11323,2,18,98,0.010421,1.5949
1433,3094,2026-02-04 16:56:21+00:00,10704,13,33,71,0.010930,1.6728
1345,2739,2025-11-06 21:01:27+00:00,9371,13,55,44,0.011952,1.8292
1284,3096,2026-02-04 17:51:29+00:00,8843,14,10,12,0.004071,0.6230
1302,3812,2026-08-10 11:19:18+00:00,8229,2,31,47,0.009722,1.4879


=== TOP-5 по ER ===


,post_id,date_time,views,reactions_total,replies,forwards,er,er_norm
1306,2827,2025-11-26 06:22:58+00:00,3044,2,307,18,0.107424,16.4408
1435,3873,2026-08-18 09:07:43+00:00,2010,4,164,18,0.092537,14.1624
1395,4051,2026-09-23 11:07:47+00:00,1446,1,116,12,0.089212,13.6535
1416,2905,2025-12-22 12:50:17+00:00,3135,8,217,28,0.080702,12.3511
1403,3277,2026-03-30 11:11:48+00:00,2243,3,147,17,0.074454,11.3949


## Мини-вывод части 1

- Данные зафиксированы в окне **12 месяцев**.
- Ориентир: сумма постов ≈ **1480**, из них ~**45** свежих (исключаются из витрин).
- Просмотры сильно варьируются → значит, есть выбросы (вирусные посты). Это важно помнить в части 3.